packages/ragbits-core/src/ragbits/core/pipelines/pipeline.py

```diff
@@ -0,0 +1,106 @@
+from abc import ABC, abstractmethod
+from types import UnionType
+from typing import Generic, TypeVar, Union, cast, get_type_hints
+
+from generics import get_filled_type
+from typing_utils import Type, issubtype
+
+InputT = TypeVar("InputT")
+OutputT = TypeVar("OutputT")
+
+
+class Step(ABC, Generic[InputT, OutputT]):
+    """
+    A step in a pipeline that processes input data and returns output data
+    """
+
+    @abstractmethod
+    async def run(self, /, input_data: InputT) -> OutputT:
+        """
+        Run the step with the given input data
+        """
+
+
+class Pipeline(Step[InputT, OutputT]):
```

Comment:
Note: Every `Pipeline` is also a `Step. Which means:

1. You can nest pipelines (add pipelines as steps of other pipelines)
2. Most likely, typing an argument as `Pipeline[input type, output type]` will never make sense - it's always better to accept any `Step[input type, output type]`. This way both `Pipeline` and `Step` can be given as the value for the argument, so users can freely choose whether they want to involve an entire pipeline or do everything in one step.

(2) also probably suggest that "Step" is not the based name. Because really `Step` can be just anything that can take some (typed) input, run some logic on it, and return some (typed) output. And `Pipeline` is just a way to chain multiple of such objects.